src/core/server/licensing/licensing_service_setup.ts

```diff
@@ -0,0 +1,191 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { i18n } from '@kbn/i18n';
+import { createHash } from 'crypto';
+import { LicenseFeature } from './license_feature';
+import { LICENSE_STATUS, LICENSE_TYPE } from './constants';
+import { LicenseType } from './types';
+
+function toLicenseType(minimumLicenseRequired: LICENSE_TYPE | string) {
+  if (typeof minimumLicenseRequired !== 'string') {
+    return minimumLicenseRequired;
+  }
+
+  if (!(minimumLicenseRequired in LICENSE_TYPE)) {
+    throw new Error(`${minimumLicenseRequired} is not a valid license type`);
+  }
+
+  return LICENSE_TYPE[minimumLicenseRequired as LicenseType];
+}
+
+export class LicensingServiceSetup {
+  private readonly _license: any;
+  private readonly license: any;
+  private readonly features: any;
+  private _signature!: string;
+  private objectified!: any;
+  private featuresMap: Map<string, LicenseFeature>;
+
+  constructor(
+    license: any,
+    features: any,
+    private error: Error | null,
+    private clusterSource: string
+  ) {
+    this._license = license;
+    this.license = license || {};
```

Comment:
I probably won't ever maintain this code, but if I did, I think I would be confused by the presence of `this._license` and `this.license`. I would be able to follow this code more easily if we only had one or the other. It looks like `this._license` is only used to assert whether there _is_ a license or not. In that case, could we do this instead?

```js
constructor(
  license: any,
  features: any,
  private error: Error | null,
  private clusterSource: string
) {
  this.hasLicense = Boolean(license);
  this.license = license || {};
```

Reply:
Agree :100: